src/AI/WallyAI.java

```diff
@@ -70,29 +51,25 @@ public AIMaker getAIInfo()
   {
     return info;
   }
-  
-  Queue<GameAction> actions = new ArrayDeque<GameAction>();
-
-  private Commander myCo = null;
 
   // What % damage I'll ignore when checking safety
   private static final int INDIRECT_THREAT_THRESHHOLD = 7;
   private static final int DIRECT_THREAT_THRESHHOLD = 13;
-  private static final double UNIT_REFUEL_THRESHHOLD = 0.25; // Fuel fraction for refuel
+  private static final int    UNIT_HEAL_THRESHHOLD = 6; // HP at which units heal
+  private static final double UNIT_REFUEL_THRESHHOLD = 1.3; // Factor of cost to get to fuel to start worrying about fuel
   private static final double UNIT_REARM_THRESHHOLD = 0.25; // Fraction of ammo in any weapon below which to consider resupply
   private static final double AGGRO_EFFECT_THRESHHOLD = 0.42; // How effective do I need to be against a unit to target it?
   private static final double AGGRO_FUNDS_WEIGHT = 0.9; // Multiplier on damage I need to get before a sacrifice is worth it
   private static final double RANGE_WEIGHT = 1; // Exponent for how powerful range is considered to be
   private static final double TERRAIN_PENALTY_WEIGHT = 3; // Exponent for how crippling we think high move costs are
   private static final double MIN_SIEGE_RANGE_WEIGHT = 0.8; // Exponent for how much to penalize siege weapon ranges for their min ranges 
-
-  private ArrayList<XYCoord> unownedProperties;
-  private ArrayList<XYCoord> capturingProperties;
   
+  private Map<UnitModel, Map<XYCoord, Double>> threatMap;
+  private ArrayList<Unit> allThreats;
   private HashMap<UnitModel, Double> unitEffectiveMove = null; // How well the unit can move, on average, on this map
   public double getEffectiveMove(UnitModel model)
```

Comment:
So I know this is unrelated to the whole modular AI breakdown, and as such should be addressed in a separate PR if at all, but why oh why are you a) recalculating each model's effective movement at the beginning of every turn, and b) looping over every map location for every model for every CO instead of just caching some numbers?

Reply:
re: first init, that's actually part of why I want to introduce persistent AI. If I fight Wally/Lash and save during a power, Wally's gonna be clinically addicted to/afraid of Recons for the rest of the match.
Then again, it might *be* worth re-evaluating partway through the match if e.g. you're fighting Ave on a map full of roads and suddenly those air units can't push as well as tanks.
...AI is hard.

As far as caching some numbers... what else would I cache?
The move cost/reachable tiles are contingent on the `MoveType`, so I can't cache between unit types.
If you mean caching the move cost per unit per terrain and then just checking that for each tile... I did. It's called a `MoveType`.

I guess I could consider only `MoveType`s, but they're 1:1 for `UnitModel`s anyway because cloning.

Reply:
At first glance, you could cache `validTiles` and `totalCosts` to avoid needing to loop over the whole map again.
That said, since this is something that only happens at init time, I'm not too worried about saving cycles.

Yeah, I hadn't really considered the whole starting mid-match thing with AI. Muriel would be affected by that as well.
Ah well. Maybe one day when we are actually good at this we'll be able to fix all the gotchas.

Reply:
I can't cache either of those values - they're unique per `MoveType`.
Sea thinks land is invalid, and land thinks sea is.

I realized the gotcha when I was setting up the saving logic, but I didn't want to open the can of worms of saving AI yet.
I think I'm happy to do that at this point since stuff has matured a bit more, though.